### PR TITLE
[VMware] Make disk controller selection on volume attachment consistent with VM creation and start

### LINKED ISSUE #9400: Failed add volume when instance running on VMware use VMware Paravirtual SCSI controller
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Instance, volume, VMware
~~~

##### CLOUDSTACK VERSION

~~~
 4.19.0.1
~~~

##### CONFIGURATION

VMware backend

##### OS / ENVIRONMENT

Ubuntu 22.04 for Management 
VMware 8.0 

##### STEPS TO REPRODUCE

~~~
Test case 1:
Set Root Disk Controller to pvscsi
Create instance from template
After create success add new volume

Test case 2:
Create Volume manual
Attach volume to instance
~~~

##### EXPECTED RESULTS

~~~
Volume attach to Instance
~~~

##### ACTUAL RESULTS

~~~
Failed to attach volume to instance

Log error:
2024-07-16 08:49:07,688 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-20:ctx-97d2789a job-22512) (logid:accc321a) Unexpected exception while executing org.apache.cloudstack.api.command.user.volume.AttachVolumeCmd
com.cloud.utils.exception.CloudRuntimeException: Failed to attach volume new-volume to VM test-vm
        at com.cloud.storage.VolumeApiServiceImpl.sendAttachVolumeCommand(VolumeApiServiceImpl.java:4402)
        at com.cloud.storage.VolumeApiServiceImpl.orchestrateAttachVolumeToVM(VolumeApiServiceImpl.java:2304)
        at com.cloud.storage.VolumeApiServiceImpl.orchestrateAttachVolumeToVM(VolumeApiServiceImpl.java:4803)
        at jdk.internal.reflect.GeneratedMethodAccessor4425.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.cloud.vm.VmWorkJobHandlerProxy.handleVmWorkJob(VmWorkJobHandlerProxy.java:107)
        at com.cloud.storage.VolumeApiServiceImpl.handleVmWorkJob(VolumeApiServiceImpl.java:4842)
        at jdk.internal.reflect.GeneratedMethodAccessor357.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
        at com.sun.proxy.$Proxy235.handleVmWorkJob(Unknown Source)
        at com.cloud.vm.VmWorkJobDispatcher.runJob(VmWorkJobDispatcher.java:102)
        at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.runInContext(AsyncJobManagerImpl.java:654)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
        at org.apache.cloudstack.managed.context.ManagedContextRunnable.run(ManagedContextRunnable.java:45)
        at org.apache.cloudstack.framework.jobs.impl.AsyncJobManagerImpl$5.run(AsyncJobManagerImpl.java:602)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
2024-07-16 08:49:07,690 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-20:ctx-97d2789a job-22512) (logid:accc321a) Complete async job-22512, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to attach volume new-volume to VM test-vm"}
~~~

### PR DESCRIPTION
**Pull request by winterhazel**
### Description

When both `rootDiskController` and `dataDiskController` details are set to SCSI controllers, the VM creation and start processes prioritize creating and using the controllers specified by `rootDiskController` for all volumes. However, the volume attachment process does not consider this and always tries to attach volumes using the controllers specified for the data disks (even if a root disk is being attached). 

This way, in a scenario where `rootDiskController` is set to `pvscsi` and `dataDiskController` is set to `lsilogic` (one example, as this problem also happens for other combinations) we ignore the `dataDiskController` setting defined by the operator and choose to use only PVSCSI controllers for both the root and data disks on VM creation/start. Then, when a user tries to attach a new data disk, we only consider `dataDiskController` and try to attach with LSI Logic instead. However, the process fails, as there are no LSI Logic controllers because we ignored the data disk controller chosen by the operator and opted to create PVSCSI controllers for both root and data disks when creating/starting the VM. Also, if we detach the root disk and try to attach it again, the process also fails because CloudStack tries to attach it using the controller specified for the data disks.

To solve this issue, this PR makes the volume attachment disk controller selection logic consistent with VM creation/start. Tests will be added in another PR alongside an extension for the disk controllers that I'm currently working on, as it will require me to modify this code again.

Fixes #9400

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [X] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

For each of the following `rootDiskController`/`dataDiskController` combinations: 
1. I deployed a VM that had 1 root disk and 1 data disk, and verified which disk controllers were being used for them;
2. I attached a data disk, and verified which disk controller was used;
3. I detached the root disk and attached it again, and verified which disk controller was used.

Below are my results.

| Configuration       | Initial root disk | Initial data disk | (re)Attached root disk | Attached data disk |
|---------------------|-------------------|-------------------|--------------------|--------------------|
| pvscsi/osdefault    | PVSCSI            | PVSCSI            | PVSCSI             | PVSCSI             |
| osdefault/pvscsi    | LSI Logic            | LSI Logic            | LSI Logic             | LSI Logic             |
| osdefault/osdefault | LSI Logic         | LSI Logic         | LSI Logic          | LSI Logic          |
| lsilogic/pvscsi     | LSI Logic         | LSI Logic         | LSI Logic          | LSI Logic          |
| pvscsi/lsilogic     | PVSCSI            | PVSCSI            | PVSCSI             | PVSCSI             |
| ide/ide             | IDE               | IDE               | IDE                | IDE                |
| ide/pvscsi          | IDE               | PVSCSI            | IDE, but see [1]   | PVSCSI             |
| lsilogic/lsilogic   | LSI Logic         | LSI Logic         | LSI Logic          | LSI Logic          |
| ide/osdefault       | IDE         | LSI Logic         | IDE, but see [1]          | LSI Logic          |
| osdefault/ide       | LSI Logic         | IDE         | LSI Logic          | IDE, but see [1]          |

[1] CloudStack tried to attach using an IDE controller. However, an exception happened because it ignored the CDRom when calculating the next available unit number, and tried to attach the disk on an unit number that was already being used. I will fix this issue separately in another PR.

### CODE REVIEW

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test ol8 vmware-70u3

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java [hunk 1]
```diff
@@ -1063,4 +1066,103 @@ public static String getAbsoluteVmdkFile(VirtualDisk disk) {
...
         return vmdkAbsFile;
     }
+
+    /**
+     * Based on an instance's <code>rootDiskController</code> and <code>dataDiskController</code> details, returns a pair
+     * containing the disk controller that should be used for root disk and the data disks, respectively.
+     */
+    public static Pair<String, String> getDiskControllersFromVmSettings(String rootDiskControllerDetail,
+                                                                        String dataDiskControllerDetail) {
+        // Validate the controller types
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/VmwareHelper.java:1076]: don´t think we need this comment, do we?
```suggestion

```

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test ol8 vmware-70u3

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test ol8 vmware-70u3

**Review by JoaoJandre — APPROVED** [review]
LGTM

**Comment by JoaoJandre** [comment]
> @winterhazel maybe some test code could be added? integration or unit. @JoaoJandre at your descretion.

Some unit testing would be nice. @winterhazel

**Comment by winterhazel** [comment]
Hey @DaanHoogland @JoaoJandre

As I stated in the description, I am planning to add tests in another PR alongside an extension for the disk controllers that I'm working on, as it will require me to make changes to both this code and to the tests again. However, if you guys prefer it, I can try adding them here already next week.

**Comment by JoaoJandre** [comment]
> Hey @DaanHoogland @JoaoJandre
> 
> As I stated in the description, I am planning to add tests in another PR alongside an extension for the disk controllers that I'm working on, as it will require me to make changes to both this code and to the tests again. However, if you guys prefer it, I can try adding them here already next week.

Sorry @winterhazel, I totally missed that part of the description.

**Pull request merged**